### PR TITLE
Shared: Add CDA parsing via QR and Paste

### PR DESCRIPTION
**Pull request by cvarley100**
# Description 

Parse CDA links scanned via QR and paste

## Type of change

- Enhancement (a non-breaking change which adds functionality)

# How Has This Been Tested?

Tested on iOS

# Checklist:

- [x] My code follows the style guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] New and existing unit tests pass locally with my changes
- [ ] For changes to `mobile` that include native code (including React Native modules): I have verified that both iOS and Android successfully build in both `Debug` and `Release` modes
- [ ] For changes to `shared`: If applicable, I have verified that my changes are implemented correctly in `desktop` and `mobile`

### CODE REVIEW

File: src/shared/actions/alerts.js [hunk 1]
```diff
@@ -47,8 +47,8 @@ const dismiss = () => ({ type: AlertsActionTypes.HIDE });
  *
  * @returns {function} dispatch
  */
-export const generateAlert = (category, title, message, closeInterval, err) => (dispatch) => {
-    dispatch(generate(category, title, message, closeInterval));
+export const generateAlert = (category, title, message, closeInterval, err, timeout) => (dispatch) => {
+    setTimeout(() => dispatch(generate(category, title, message, closeInterval)), timeout);
```
Comment (rajivshah3) [inline src/shared/actions/alerts.js]:
```suggestion
    if (!timeout) {
        dispatch(generate(category, title, message, closeInterval));
    } else {
        setTimeout(() => dispatch(generate(category, title, message, closeInterval)), timeout);
    }
```

If `generateAlert` is called and `timeout` is not specified, a timer will be created with a timeout of 0. On mobile, this actually results in a [call](https://github.com/facebook/react-native/blob/c20070f10458d48d6ac1eaac49e681e932bfb9fd/Libraries/Core/Timers/JSTimers.js#L259) to the native timing module. Even though the native timing module will immediately call the callback if `timeout` is 0 (see [iOS](https://github.com/facebook/react-native/blob/c20070f10458d48d6ac1eaac49e681e932bfb9fd/React/Modules/RCTTiming.m#L280-L284) and [Android](https://github.com/facebook/react-native/blob/c20070f10458d48d6ac1eaac49e681e932bfb9fd/ReactAndroid/src/main/java/com/facebook/react/modules/core/Timing.java#L347-L352)), using `setTimeout` for something that will run immediately is still unnecessary IMHO. This change will also resolve most of the failing tests for `shared`.

File: src/shared/libs/iota/utils.js [hunk 2]
```diff
@@ -546,3 +546,52 @@ export const withRequestTimeoutsHandler = (timeout) => {
...
+        message: undefined,
+        expectedAmount: undefined,
+        timeoutAt: undefined,
+    };
+
+    try {
+        const url = new URL(input, true)
+        const { address, message, amount, timeoutAt } = url.query
+
+        if (address.match(/^[A-Z9]{90}$/)) {
```
Comment (rajivshah3) [inline src/shared/libs/iota/utils.js]:
```suggestion
        if (address.match(VALID_ADDRESS_WITH_CHECKSUM_REGEX)) {
```

File: src/shared/libs/iota/utils.js [hunk 3]
```diff
@@ -546,3 +546,52 @@ export const withRequestTimeoutsHandler = (timeout) => {
...
+            result.address = address.toUpperCase();
```
Comment (rajivshah3) [inline src/shared/libs/iota/utils.js]:
```suggestion
            result.address = address;
```

If the address matches the above regex, shouldn't it already be upper case?

**Review by rajivshah3 — CHANGES_REQUESTED** [review]

File: src/desktop/src/ui/global/Alerts.js [hunk 4]
```diff
@@ -47,7 +47,7 @@ export class AlertsComponent extends React.PureComponent {
         if (nextProps.alerts.category && nextProps.alerts.category.length && nextProps.alerts.closeInterval > 0) {
             this.timeout = setTimeout(() => {
                 this.props.dismissAlert();
-            }, nextProps.alerts.closeInterval);
+            }, nextProps.alerts.closeInterval || 0);
```
Comment (rajivshah3) [inline src/desktop/src/ui/global/Alerts.js:50]:
```suggestion
            }, nextProps.alerts.closeInterval);
```
If `closeInterval > 0` (in the `if` statement), shouldn't `closeInterval || 0` always evaluate to `closeInterval`?

File: src/shared/actions/alerts.js (hunk 1 above)
Comment (cvarley100) [reply src/shared/actions/alerts.js]: Okay good catch.

File: src/desktop/src/ui/global/Alerts.js (hunk 4 above)
Comment (cvarley100) [reply src/desktop/src/ui/global/Alerts.js:50]: Could be undefined

File: src/shared/actions/transfers.js [hunk 5]
```diff
@@ -1022,3 +1025,46 @@ export const retryFailedTransaction = (accountName, bundleHash, seedStore, quoru
...
+          dispatch(setCDAContent(data));
+          dispatch(setSendAddressField(data.address));
+          if (data.expectedAmount) {
+              dispatch(setSendAmountField(data.expectedAmount.toString()));
+          }
+          if (data.message) {
+              dispatch(setSendMessageField(data.message));
+          }
+      } catch (err) {
+          if (err.message === 'Expired timeout.') {
```
Comment (laumair) [inline src/shared/actions/transfers.js]: `Expired timeout.` should probably go to https://github.com/iotaledger/trinity-wallet/blob/feat/exchanges/src/shared/libs/errors.js

File: src/desktop/src/ui/views/wallet/Dashboard.js [hunk 6]
```diff
@@ -67,6 +79,16 @@ class Dashboard extends React.PureComponent {
         );
     };
 
+    onPaste = (e) => {
+        const data = (event.clipboardData || window.clipboardData).getData('text')
+        const parsedCDALink = parseCDALink(data);
+        if (parsedCDALink){
+            this.props.verifyCDAContent(parsedCDALink);
+            this.props.history.push('/wallet/send');
+            return e.preventDefault();
```
Comment (laumair) [inline src/desktop/src/ui/views/wallet/Dashboard.js:88]: Shouldn't we _always_ prevent the default event? As in shouldn't `e.preventDefault` always be called when `onPaste` callback is triggered?

Comment (cvarley100) [reply src/desktop/src/ui/views/wallet/Dashboard.js:88]: No, because we want it to fall through to the regular paste function(s) otherwise.

**Review by laumair — APPROVED** [review]

**Pull request merged**